### ISSUE TITLE
False positives in the disposable ownership rules

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/eslint-plugin#87]
The `require-disposable-ownership` and `require-disposable-transfer` rules added in #77 produce 16 findings on the JupyterLab downstream run. 14 of those are false positives. This issue covers the four causes that are straightforward to fix. Each was confirmed against the JupyterLab source and reproduced as a minimal test case.

| Cause | Downstream findings | Should be fixed before release? |
| --- | --- | --- |
| Exported module singleton reported as unowned | 2 | yes |
| Plugin `activate` written as a named function is not recognised | 5 | yes |
| Disposal inside a callback is not recognised | 2 | yes |
| `ignoredReturnFunctionNames` silently widens the detection scope | 0 (API bug) | yes |
| Ownership through an intermediate options variable | 3 | no, follow-up |
| Ownership handed to a local helper function | 1 | no, follow-up |
| Constructor options map is missing an entry | 1 | no, follow-up |

The two remaining findings are genuine: `packages/toc/src/factory.ts:77` creates an `ActivityMonitor` that is never disposed, and `packages/apputils/src/licenses.tsx:131` creates a `Widget` purely to obtain its `title`.

### 1. Exported module singletons are reported as unowned

Both rules report a disposable assigned to an exported module level binding.

Downstream findings:

- `packages/apputils/src/dialog.tsx:1127`, `export const tracker = new WidgetTracker<Dialog<any>>({ ... })` inside `export namespace Dialog`
- `packages/apputils/src/notification.ts:398`, `export const manager = new NotificationManager()` inside `export namespace Notification`

Reproduction:

```ts
declare class DisposableDelegate { constructor(fn: () => void); }
export const tracker = new DisposableDelegate(() => undefined);
```

An exported binding is created exactly once and lives for the lifetime of the program, so there is nothing to leak. Its disposal, if any is wanted at all, is the responsibility of the importers, which the rule cannot see from this file. Asking the declaring module to own it is not actionable.

Cause: `getAssignedVariable` resolves the variable and the only lifetime guard afterwards is `isOuterFunctionScopeVariable`, which compares the function scope of the creation site against the function scope of the variable. For a top level declaration both are the module scope, so they match and the variable is treated exactly like a function local.

Restricting the exemption to exported bindings clears both downstream findings and keeps all of those tests passing. A broader "any module scope variable" exemption would break 17 existing test cases. 

### 2. Plugin `activate` written as a named function is not recognised

Both rules exempt disposables created directly inside a plugin `activate` function, on the grounds that plugin services usually live for the application lifetime. The exemption only matches one of the three ways `activate` is written, so it misses the most common one in larger extensions.

Downstream findings, all of them disposables created directly in the body of an activation function:

- `packages/apputils-extension/src/palette.ts:98`, `modalPalette` in `Private.activate`
- `packages/csvviewer-extension/src/index.ts:144` and `:311`, `tracker` in `activateCsv` and `activateTsv`
- `packages/toc-extension/src/index.ts:290`, `toolbarMenu` in `activateTOC`
- `packages/console-extension/src/index.ts:340`, `promptMenu` in `activateConsole`

Reproduction:

```ts
declare class DisposableDelegate { constructor(fn: () => void); }
function activateCsv(): void {
  const tracker = new DisposableDelegate(() => undefined);
  void tracker;
}
const plugin = { id: 'x:plugin', autoStart: true, activate: activateCsv };
export default plugin;
```

Cause: `isInJupyterPluginActivate` requires the enclosing function to be the `value` of a `Property` named `activate` on a plugin shaped object literal. That covers `activate: () => { ... }` written inline. It does not cover a separate function referenced as `activate: activateCsv`, and it does not cover `export function activate()` in a `Private` namespace.

### 3. Disposal inside a callback is not recognised

Both rules report a disposable that is disposed inside a callback rather than in the same statement block where it was created.

Downstream findings:

- `packages/apputils/src/thememanager.ts:523`, `splash` is disposed at `:571` inside `requestAnimationFrame(() => { ... })` and at `:576` inside `.catch(reason => { ... })`
- `packages/apputils/src/toolbar/factory.ts:260`, `items` is captured by closures created in `createToolbarFactory` and used after the function returns

Reproduction:

```ts
declare class DisposableDelegate { constructor(fn: () => void); dispose(): void; }
declare function requestAnimationFrame(cb: () => void): void;
function load(): void {
  const splash = new DisposableDelegate(() => undefined);
  requestAnimationFrame(() => {
    splash.dispose();
  });
}
```

Cause: ownership is only credited through `isManagedUse`, which is `isUnconditionalUse` or `isTruthyGuardedUse`. Both stop walking at the first function boundary, so any use inside a callback is invisible. The file already has a narrow workaround for this in `markDisposedSignalCallbackDisposals`, but it is gated to callbacks passed to `<something>.disposed.connect(...)` specifically.

**This in conflict with test cases that were added on purpose**. They assert that this is an error, because `defer` might never run:

```ts
const disposable = createDisposable();
defer(() => {
  disposable.dispose();
});
```

That is structurally identical to `thememanager.ts:568-572`, which is correct code. There is no way to accept one and reject the other, so this is a policy choice rather than a bug.

We can take the position that disposal inside a callback is disposal, on the grounds that a rule shipped at `warn` in the recommended config should prefer a missed detection over a false positive on correct code. To limit what is given up, the fix could only accept unconditional disposal inside the callback, which preserves two of the four affected cases (the ones where the disposal sits inside an `if` within the callback).

### 4. `ignoredReturnFunctionNames` silently widens the detection scope

In `require-disposable-transfer`, setting the `ignoredReturnFunctionNames` option changes far more than the ignore list. Three separate behaviours are gated on whether the option was supplied:

- the exemption for `this._map.set(...)` and `this._map.delete(...)`
- the exemption for names matching `add*Factory`
- the restriction that only factory named calls (`create*`, `build*`, `make*`, `new*`) have their return value checked at all

The first two are defensible, since they are part of the default exemptions and replacing the default list can reasonably replace them too. The third is a detection scope change, not an ignore list change. A user who adds a single name to the ignore list to silence one call site instead gets reports on every call in their codebase whose return type happens to be disposable. Nothing in the rule documentation mentions this.
